tomcat-8.5/src/org/dogtagpki/tomcat/JSSContext.java

```diff
@@ -0,0 +1,102 @@
+package org.dogtagpki.tomcat;
+
+import java.util.List;
+import java.security.Provider;
+import java.security.Security;
+
+import javax.net.ssl.KeyManager;
+import javax.net.ssl.KeyManagerFactory;
+
+import org.apache.tomcat.util.net.SSLContext;
+
+import org.mozilla.jss.provider.javax.crypto.JSSKeyManager;
+import org.mozilla.jss.ssl.javax.JSSEngine;
+import org.mozilla.jss.ssl.javax.JSSParameters;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JSSContext implements org.apache.tomcat.util.net.SSLContext {
+    public static Logger logger = LoggerFactory.getLogger(JSSContext.class);
+
+    private javax.net.ssl.SSLContext ctx;
+    private JSSParameters params;
+
+    public JSSContext(String alias, List<String> negotiableProtocols) {
+        logger.debug("TomcatJSSContext(" + alias + ", ...");
+
+        params = new JSSParameters();
+        params.setAlias(alias);
+        params.setProtocols(negotiableProtocols.toArray(new String[0]));
+    }
+
+    public void init(javax.net.ssl.KeyManager[] kms, javax.net.ssl.TrustManager[] tms, java.security.SecureRandom sr) throws java.security.KeyManagementException {
+        logger.debug("TomcatJSSContext.init(...)");
+        try {
+            ctx = javax.net.ssl.SSLContext.getDefault();
+        } catch (java.security.NoSuchAlgorithmException e) {
+            throw new java.security.KeyManagementException(e.getMessage(), e);
+        }
+        ctx.init(kms, tms, sr);
+    }
+
+    public javax.net.ssl.SSLEngine createSSLEngine() {
+        logger.debug("TomcatJSSContext.createSSLEngine()");
+        JSSEngine eng = (JSSEngine) ctx.createSSLEngine();
+        eng.setSSLParameters(params);
+        return eng;
+    }
+
+    public javax.net.ssl.SSLSessionContext getServerSessionContext() {
+        logger.debug("TomcatJSSContext.getServerSessionContext()");
+        return ctx.getServerSessionContext();
+    }
+
+    public javax.net.ssl.SSLServerSocketFactory getServerSocketFactory() {
+        logger.debug("TomcatJSSContext.getServerSocketFactory()");
+        return ctx.getServerSocketFactory();
+    }
+
+    public javax.net.ssl.SSLParameters getSupportedSSLParameters() {
+        logger.debug("TomcatJSSContext.getSupportedSSLParameters()");
+        return params;
+    }
+
+    public java.security.cert.X509Certificate[] getCertificateChain(java.lang.String alias) {
+        logger.debug("TomcatJSSContext.getCertificateChain(" + alias + ")");
+
+        Exception last = null;
+        KeyManagerFactory kmf = null;
+
+        try {
+            kmf = KeyManagerFactory.getInstance("NssX509");
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to load NSS KeyManagerFactory: " + e.getMessage(), e);
+        }
+
+        for (KeyManager km : kmf.getKeyManagers()) {
```

Comment:
I think we should use the KeyManagers provided during initialization, and we need to make sure that we configure it with JSS key managers.

Reply:
See discussion in your first comment for context. I'm thinking I should add the missing methods to `JSSContext` so that we can simply call into it from this one for *all* methods. 

The issue here is that I don't want to loop to find a `JSSKeyManager` twice (which we already do in JSS's `JSSContext`). I would have to make the internal `JSSKeyManager` instance `public` if I wanted to access it from over here in Tomcat. 

I think if we add this method to `org.mozilla.jss.provider.javax.net.JSSContext`, we can use the `JSSKeyManager` we already found at initialization and simplify this method. 

Thoughts?

Reply:
It isn't possible to implement both in the same class. The reason is that `SSLContextSpi` is meant to be called from the `SSLContext` (final class). Providers implement the Spi interface and cannot add additional methods: `SSLContext` simply doesn't know how to handle them. 

So, we can't merge these two classes how I'd like to.